java/ql/lib/semmle/code/java/security/regexp/RegexInjection.qll

```diff
@@ -0,0 +1,51 @@
+/** Provides classes and predicates related to regex injection in Java. */
+
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.frameworks.Regex
+private import semmle.code.java.regex.RegexFlowModels
+
+/** A data flow sink for untrusted user input used to construct regular expressions. */
+abstract class RegexInjectionSink extends DataFlow::ExprNode { }
+
+/** A sanitizer for untrusted user input used to construct regular expressions. */
+abstract class RegexInjectionSanitizer extends DataFlow::ExprNode { }
+
+/** A method call that takes a regular expression as an argument. */
+private class DefaultRegexInjectionSink extends RegexInjectionSink {
+  DefaultRegexInjectionSink() {
+    exists(string kind |
+      kind.matches(["regex-use[]", "regex-use[f1]", "regex-use[f-1]", "regex-use[-1]", "regex-use"]) and
+      sinkNode(this, kind)
+    )
```

Comment:
I you're not using wildcards (still curious why `regex-use%` doesn't work), you don't need `matches` (and then the `exists` isn't really needed either):

```suggestion
    sinkNode(this, ["regex-use[]", "regex-use[f1]", "regex-use[f-1]", "regex-use[-1]", "regex-use"])
```

Reply:
Some of the regex-use sinks are specific to the ReDoS query; in particular the `Predicate.test` sink is likely to cause many false positives if included. Which is probably the reason regex-use[0] is being excluded here. 

The best solution is probably to exclude that sink specifically - perhaps by giving it a seperate kind that only the ReDoS query checks for.

Reply:
Ah, makes sense, thanks! I would be surprised if more queries used these sinks aside from ReDos and regex injection, but if you think it makes sense we can create a special sink kind for `Predicate.test`. Specially if `regex-use[0]` includes other sinks that are valid for this query.

Reply:
Updated in https://github.com/github/codeql/pull/11070/commits/13decd38d90d30449f0a875278c35d8a2d9f3772, and made a follow-up issue as discussed.